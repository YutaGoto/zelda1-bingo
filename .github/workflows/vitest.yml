name: vitest

on:
  - push

jobs:
  vitest:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout your repository using git
        uses: actions/checkout@v4
      - name: Setup pnpm
        uses: pnpm/action-setup@v2
        with:
          version: 8
      - name: Setup Node.js
        uses: actions/setup-node@v4
        with:
          cache: "pnpm"
          node-version: 20
      - name: Install dependencies
        run: pnpm install
      - name: Run tests
        run: pnpm test
