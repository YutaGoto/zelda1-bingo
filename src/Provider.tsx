import {
  ChakraProvider,
  ColorModeScript,
  type ThemeConfig,
  extendTheme,
} from "@chakra-ui/react";
import { type StyleFunctionProps, mode } from "@chakra-ui/theme-tools";
import type { ReactNode } from "react";

interface ProviderProps {
  children: ReactNode;
}

export const Provider = ({ children }: ProviderProps) => {
  const config: ThemeConfig = {
    initialColorMode: "system",
    useSystemColorMode: true,
  };

  const theme = extendTheme({
    ...config,
    styles: {
      global: (props: StyleFunctionProps) => ({
        body: {
          bg: mode("gray.50", "gray.800")(props),
        },
      }),
    },
    fonts: {
      heading: `'Noto Sans JP Variable', sans-serif`,
      body: `'Noto Sans JP Variable', sans-serif`,
    },
  });

  return (
    <ChakraProvider theme={theme}>
      <ColorModeScript initialColorMode={config.initialColorMode} />
      {children}
    </ChakraProvider>
  );
};
