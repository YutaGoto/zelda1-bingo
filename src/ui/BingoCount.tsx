import { Text } from "@chakra-ui/react";
import { useEffect, useState } from "react";
import { useTranslation } from "react-i18next";

interface BingoCountProps {
  hits: boolean[][];
}

export const BingoCount = ({ hits }: BingoCountProps) => {
  const { t } = useTranslation();
  const [bingoCount, setBingoCount] = useState<number>(0);

  useEffect(() => {
    let newBingoCount = hits.reduce((acc, row) => {
      if (row.every((hit) => hit)) {
        return acc + 1;
      }

      return acc;
    }, 0);

    newBingoCount += hits.reduce((acc, _row, i) => {
      return acc + (hits.map((row) => row[i]).every((hit) => hit) ? 1 : 0);
    }, 0);

    if (hits[0][0] && hits[1][1] && hits[2][2] && hits[3][3] && hits[4][4]) {
      newBingoCount += 1;
    }

    if (hits[0][4] && hits[1][3] && hits[2][2] && hits[3][1] && hits[4][0]) {
      newBingoCount += 1;
    }

    setBingoCount(newBingoCount);
  }, [hits]);

  return (
    <Text fontSize="2xl" className="bingo-count">
      {t("bingoCount", { count: bingoCount })}
    </Text>
  );
};
