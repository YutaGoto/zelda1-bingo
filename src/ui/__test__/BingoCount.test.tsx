import { render, screen } from "@testing-library/react";
import { expect, test } from "vitest";

import { Provider } from "../../Provider";
import { BingoCount } from "../BingoCount";

test("BingoCount", () => {
  render(
    <BingoCount
      hits={[
        [true, false, false, false, false],
        [false, true, false, false, false],
        [false, false, true, false, false],
        [false, false, false, true, false],
        [false, false, false, false, true],
      ]}
    />,
    { wrapper: Provider },
  );

  expect(screen.getByText("1 line Bingo")).toBeTruthy();
});

test("Snapshot", () => {
  const { asFragment } = render(
    <BingoCount
      hits={[
        [true, false, false, false, false],
        [false, true, false, false, false],
        [false, false, true, false, false],
        [false, false, false, true, false],
        [false, false, false, false, true],
      ]}
    />,
    { wrapper: Provider },
  );

  expect(asFragment()).toMatchSnapshot();
});
