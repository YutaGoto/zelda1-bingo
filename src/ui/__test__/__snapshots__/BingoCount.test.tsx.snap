// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshot 1`] = `
<DocumentFragment>
  <p
    class="chakra-text bingo-count css-1n49pl7"
  >
    1 line Bingo
  </p>
</DocumentFragment>
`;
